linkerd/stack/src/gate.rs

```diff
@@ -145,117 +163,210 @@ where
     type Future = S::Future;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        // If the gate is shut, wait for it to open by storing a future that
-        // will complete when it opens. This ensures that waiters are notified.
-        while self.rx.is_shut() {
-            trace!(gate.open = false);
-            if !self.is_waiting {
-                let rx = self.rx.clone();
-                self.waiting.set(async move { rx.opened().await });
-                self.is_waiting = true;
+        let permit = ready!(self.poll_acquire(cx));
+        ready!(self.inner.poll_ready(cx))?;
+        self.permit = Poll::Ready(Permit(permit));
+        Poll::Ready(Ok(()))
+    }
+
+    #[inline]
+    fn call(&mut self, req: Req) -> Self::Future {
+        self.permit = match std::mem::replace(&mut self.permit, Poll::Pending) {
+            Poll::Pending => panic!("poll_ready must be called first"),
+            Poll::Ready(..) => Poll::Pending,
+        };
```

Comment:
not sure if i understand the match and reassignment to `self.permit` here --- the `mem::replace` call should already have set `self.permit` to `Poll::Pending`. isn't this just
```suggestion
        if std::mem::replace(&mut self.permit, Poll::Pending).is_pending() {
            panic!("poll_ready must be called first");
        }
```
or even
```suggestion
        let permit = std::mem::replace(&mut self.permit, Poll::Pending);
        assert!(permit.is_ready(), "poll_ready must be called first");
```